### PR TITLE
Implement SSIDDI

### LINKED ISSUE #11: Add the SSI-DDI Model
- Please read the paper first. It is [here](https://academic.oup.com/bib/article-abstract/22/6/bbab133/6265181?redirectedFrom=fulltext).
- There is also code-release with the paper [here](https://github.com/kanz76/SSI-DDI).
- After that read the contributing guidelines.
-   If there is an existing open-source version of the model please take a look.
-   ChemicalX is built on top of [PyTorch 1.10.](https://pytorch.org/) and [torchdrug](https://torchdrug.ai/).
- A similar model is `EPGCNDS` which uses`GraphConvolutions`  to generate drug representations. Take a look at the layer definition [here](https://github.com/AstraZeneca/chemicalx/blob/main/chemicalx/models/epgcnds.py). You should use the layers from `torchdrug` not the models.
- The library heavily builds on top on [torchdrug](https://torchdrug.ai/) and molecules in batches are PackedGraphs.
- There is already a model class under `./chemicalx/models/`
- Context features, drug level features, and labels are all FloatTensors.
- Look at the examples and tests under `./examples/` and `./tests/`.
- Add auxiliary layers as you see fit - please document these, add tests, and add these layers to the main readme.md if needed.
- Add typing to the initialization and forward pass.
- Non-data-dependent hyper-ammeters should have default values. 
- Please add tests under `./tests/` and make sure that your model/layer is tested with real data.
- Write an example under `./examples/`. What is the AUC on the test set? Is it reasonable?

### PR DESCRIPTION
**Pull request by YuWVandy**
Closes #11

# Summary
- [1] Add the SSIDDI model and all its components to ssiddi.py

### CODE REVIEW

File: chemicalx/models/ssiddi.py [hunk 1]
```diff
@@ -1,11 +1,318 @@
...
+            i * max_num_nodes for i in range(batch_size)
+        ]
+        mask = torch.cat(mask, dim=0)
+
+        perm = perm[mask]
+
+    return perm
+
+
+def softmax(src, batch):
```
Comment (cthoyt) [inline chemicalx/models/ssiddi.py:85]: is this different from https://pytorch-scatter.readthedocs.io/en/1.4.0/composite/softmax.html#torch_scatter.composite.scatter_softmax?

Comment (YuWVandy) [reply chemicalx/models/ssiddi.py:85]: No, they are the same. I think we can directly use this package here. Thanks for the check.

File: chemicalx/models/ssiddi.py [hunk 2]
```diff
@@ -1,11 +1,318 @@
...
+from chemicalx.data import DrugPairBatch
+from chemicalx.models import Model
+import torch.nn.functional as F
+
 
 __all__ = [
     "SSIDDI",
 ]
 
+def filter_adj(edge_index, edge_attr, perm, num_nodes):
```
Comment (benedekrozemberczki) [inline chemicalx/models/ssiddi.py:24]: Please add typing.

Comment (benedekrozemberczki) [reply chemicalx/models/ssiddi.py:24]: Also add the docstring.

File: chemicalx/models/ssiddi.py [hunk 3]
```diff
@@ -1,11 +1,318 @@
...
+    mask = (row >= 0) & (col >= 0)
+    row, col = row[mask], col[mask]
+
+    if edge_attr is not None:
+        edge_attr = edge_attr[mask]
+
+    return torch.stack([row, col], dim=0), edge_attr
+
+
+def topk(x, ratio, batch, min_score=None, tol=1e-7):
```
Comment (benedekrozemberczki) [inline chemicalx/models/ssiddi.py:40]: Add typing and docstring.

File: chemicalx/models/ssiddi.py [hunk 4]
```diff
@@ -1,11 +1,318 @@
...
+
+    if(edge_weight == None):
+        edge_weight = torch.ones_like(row)
+
+    out = edge_weight.view(-1, 1) * x[row]
+
+    return scatter(out, col, dim=0, dim_size=x.size(0), reduce='add')
+
+
+class GraphConv(torch.nn.Module):
```
Comment (benedekrozemberczki) [inline chemicalx/models/ssiddi.py:108]: There is a dedicated GCN layer in torhdrug.

**Review by benedekrozemberczki — CHANGES_REQUESTED** [review]
Can you please add tests, use the torchdrug graph convolution, add typind and docstrings everywhere.

File: chemicalx/models/ssiddi.py (hunk 4 above)
Comment (YuWVandy) [reply chemicalx/models/ssiddi.py:108]: A little bit different from GCN. Here they actually use the Conv layer from here https://arxiv.org/abs/1810.02244.

**Comment by benedekrozemberczki** [comment]
@cthoyt and @YuWVandy Refactored this one, added tests, documentation etc.

**Review by benedekrozemberczki — APPROVED** [review]
